library/sql.inc

```diff
@@ -514,19 +514,31 @@ function sqlNumRows($r)
 */
 function HelpfulDie ($statement, $sqlerr='')
 {
-  echo "<p><p><font color='red'>ERROR:</font> ".text($statement)."<p>";
+
+  if($GLOBALS['sql_string_no_show_screen']){
+    echo "<h2>" . xlt('Save Unsuccessful.') . "</h2>";
+  }
+
+  if(!$GLOBALS['sql_string_no_show_screen']) {
+    echo "<p><p><font color='red'>ERROR:</font> " . text($statement) . "<p>";
+  }
   $logMsg="SQL Error with statement:".$statement;
   if ($sqlerr) {
-    echo "Error: <font color='red'>".text($sqlerr)."</font><p>";
+    if(!$GLOBALS['sql_string_no_show_screen']) {
+       echo "Error: <font color='red'>" . text($sqlerr) . "</font><p>";
+    }
     $logMsg.="--".$sqlerr;
   }//if error
-  $backtrace=debug_backtrace();
-  for($level=1;$level<count($backtrace);$level++)
-  {
-      $info=$backtrace[$level];
-      echo "<br>".text($info["file"]." at ".$info["line"].":".$info["function"]);
-      if($level>1){
-          echo "(".text(implode(",",$info["args"])).")";
+
+  if(!$GLOBALS['sql_string_no_show_screen']) {
+      $backtrace = debug_backtrace();
+      for ($level = 1; $level < count($backtrace); $level++)
+      {
+          $info = $backtrace[$level];
+          echo "<br>" . text($info["file"] . " at " . $info["line"] . ":" . $info["function"]);
+          if ($level > 1) {
+              echo "(" . text(implode(",", $info["args"])) . ")";
+          }
       }
   }
```

Comment:
migrate the $backtrace = debug_backtrace(); to happen prior to this if (this is because the $backtrace is used below to send error to php error log.